aptos-move/aptos-aggregator/src/aggregator_change_set.rs

```diff
@@ -2,331 +2,447 @@
-                AggregatorChange::StringData { .. } => Err(code_invariant_error(
-                    "Base value cannot be string for an aggregator",
-                )),
-                AggregatorChange::AggregatorDelta {
-                    delta: prev_delta,
-                    max_value: prev_max_value,
-                    history: prev_history,
-                } => {
-                    if max_value != prev_max_value {
-                        return Err(code_invariant_error(
-                            "Max value of aggregator cannot change",
-                        ));
-                    }
-                    let (new_delta, new_history) =
-                        merge_two_deltas(prev_delta, prev_history, delta, history, *max_value)?;
-                    *self = AggregatorChange::AggregatorDelta {
-                        delta: new_delta,
-                        history: new_history,
-                        max_value: *max_value,
-                    };
-                    Ok(())
-                },
-                AggregatorChange::SnapshotDelta { .. } => Err(code_invariant_error(
-                    "SnapshotDelta cannot come before AggregatorDelta",
-                )),
+    pub fn get_apply_base_id(&self, self_id: &AggregatorID) -> ApplyBase {
+        self.get_apply_base_id_option()
+            .unwrap_or(ApplyBase::Previous(*self_id))
+    }
+
+    pub fn apply_to_base(&self, base_value: AggregatorValue) -> PartialVMResult<AggregatorValue> {
+        use AggregatorApplyChange::*;
+
+        Ok(match self {
+            AggregatorDelta { delta } => {
+                AggregatorValue::Aggregator(delta.apply_to(base_value.into_aggregator_value()?)?)
             },
-            AggregatorChange::SnapshotDelta { .. } => {
-                unimplemented!()
+            SnapshotDelta { delta, .. } => {
+                AggregatorValue::Snapshot(delta.apply_to(base_value.into_aggregator_value()?)?)
             },
+            SnapshotDerived { formula, .. } => {
+                AggregatorValue::Derived(formula.apply(base_value.into_snapshot_value()?))
+            },
+        })
+    }
+}
+
+impl AggregatorChange {
+    // When squashing a new change on top of the old one, sometimes we need to know the change
+    // from a different AggregatorID to be able to merge them together.
+    // In particular SnapshotDelta represents a change from the aggregator at the beginning of the transaction,
+    // and squashing changes where the aggregator will be at the beginning of the transaction.
+    // For example, let’s say we have two change sets that we need to squash:
+    // change1: agg1 -> Delta(+3)
+    // change2: agg1 -> Delta(+6), snap1 -> (base=agg1, Delta(+2))
+    // the correct squashing of snapshot depends on the change for the base aggregator. I.e. the correct output would be:
+    // agg1 -> Delta(+9), snap(base=agg1, Delta(+5))
+    pub fn get_merge_dependent_id(&self) -> Option<AggregatorID> {
```

Comment:
Shouldn't this function be part of `impl AggregatorApplyChange` instead?

Reply:
we are merging AggregatorChange in change_set.rs. I don't want to have there a match on whether a change is a value or not. and it also corresponds to merge_two_changes function from AggregatorChange - so keeping it there makes sense logically